File: src/karabo/util/TimeDuration.cc
/* 
 * File:   TimeDuration.cc
 * Author: boukhelef
 * 
 * Created on April 27, 2013, 10:55 PM
 */

#include "TimeDuration.hh"

namespace karabo {
    namespace util {

        std::string TimeDuration::DEFAULT_FORMAT("%s.%N");


        TimeDuration::TimeDuration() :
            m_Seconds(0ULL),
            m_Fractions(0ULL) {
        }


        TimeDuration::TimeDuration(const karabo::util::Hash& hash) {
            fromHash(hash);
        }


        TimeDuration::TimeDuration(const TimeValue seconds, const TimeValue fractions) :
            m_Seconds(seconds),
            m_Fractions(fractions) {
            sanitize(m_Seconds, m_Fractions);
        }


        TimeDuration::TimeDuration(const int days, const int hours, const int minutes, const TimeValue seconds, const TimeValue fractions) :
            m_Seconds(days*DAY + hours*HOUR + minutes*MINUTE + seconds),
            m_Fractions(fractions) {
            sanitize(m_Seconds, m_Fractions);
        }


        TimeDuration::~TimeDuration() {
        }


        TimeDuration& TimeDuration::set(const TimeValue seconds, const TimeValue fractions) {
            m_Seconds = seconds;
            m_Fractions = fractions;
            sanitize(m_Seconds, m_Fractions);
            return *this;
        }


        TimeDuration & TimeDuration::set(const int days, const int hours, const int minutes, const TimeValue seconds, const TimeValue fractions) {
            m_Seconds = days * DAY + hours * HOUR + minutes * MINUTE + seconds;
            m_Fractions = fractions;
            sanitize(m_Seconds, m_Fractions);
            return *this;
        }


        TimeDuration& TimeDuration::add(TimeValue seconds, TimeValue fractions) {
            sanitize(seconds, fractions);
            m_Seconds += seconds;
            m_Fractions += fractions;
            if (m_Fractions > m_oneSecondInAtto) {
                ++m_Seconds;
                m_Fractions -= m_oneSecondInAtto;
            }
            return *this;
        }


        TimeDuration & TimeDuration::add(const int days, const int hours, const int minutes, const TimeValue seconds, const TimeValue fractions) {
            m_Seconds += days * DAY + hours * HOUR + minutes * MINUTE + seconds;
            m_Fractions += fractions;
            sanitize(m_Seconds, m_Fractions);
            return *this;
        }


        TimeDuration& TimeDuration::sub(const TimeValue seconds, const TimeValue fractions) {
            m_Seconds -= seconds;
            if (m_Fractions < fractions) {
                m_Fractions = (m_oneSecondInAtto - fractions) + m_Fractions;
                --m_Seconds;
            } else {
                m_Fractions -= fractions;
            }
            return *this;
        }


        TimeDuration & TimeDuration::sub(const int days, const int hours, const int minutes, const TimeValue seconds, const TimeValue fractions) {
            m_Seconds -= days * DAY + hours * HOUR + minutes * MINUTE + seconds;
            if (m_Fractions < fractions) {
                m_Fractions = (m_oneSecondInAtto - fractions) + m_Fractions;
                --m_Seconds;
            } else {
                m_Fractions -= fractions;
            }
            return *this;
        }


        bool TimeDuration::isNull() const {
            return (m_Seconds == 0ull) && (m_Fractions == 0ull);
        }


        uint64_t TimeDuration::getDays() const {
            return m_Seconds / DAY;
        }


        uint64_t TimeDuration::getHours() const {
            return (m_Seconds / HOUR) % 24;

        }


        TimeValue TimeDuration::getTotalHours() const {
            return m_Seconds / HOUR;
        }


        uint64_t TimeDuration::getMinutes() const {
            return (m_Seconds / MINUTE) % 60;
        }


        TimeValue TimeDuration::getTotalMinutes() const {
            return m_Seconds / MINUTE;
        }


        TimeValue TimeDuration::getSeconds() const {
            return m_Seconds % 60;
        }


        TimeValue TimeDuration::getTotalSeconds() const {
            return m_Seconds;
        }


        TimeValue TimeDuration::getFractions(const TIME_UNITS unit) const {
            int zeros = int(unit);
            unsigned long long multiplier = 1ULL;
            while (zeros-- > 0) multiplier *= 10ULL;
            return m_Fractions / multiplier; // unsigned long long dividing
        }


        std::string TimeDuration::format(const std::string& fmt) const {
            std::ostringstream oss;
            oss.fill('0');
            for (char const * ptr = fmt.c_str(); *ptr; ++ptr) {
                if (*ptr == '%') {
                    ++ptr;
                    switch (*ptr) {
                        case 'd':
                        case 'D': oss << getDays();
                            break;
                        case 'H': oss.width(2);
                            oss << getHours();
                            break;
                        case 'M':oss.width(2);
                            oss << getMinutes();
                            break;
                        case 'S': oss.width(2);
                            oss << getSeconds();
                            break;
                        case 'h': oss.width(0);
                            oss << getHours();
                            break;
                        case 'm': oss.width(0);
                            oss << getMinutes();
                            break;
                        case 's': oss.width(0);
                            oss << getSeconds();
                            break;
                        default:
                        {
                            int w;
                            TIME_UNITS p = ATTOSEC;
                            switch (*ptr) {
                                case 'l': w = 0, p = MILLISEC;
                                    break;
                                case 'u': w = 0, p = MICROSEC;
                                    break;
                                case 'n': w = 0, p = NANOSEC;
                                    break;
                                case 'p': w = 0, p = PICOSEC;
                                    break;
                                case 'f': w = 0, p = FEMTOSEC;
                                    break;
                                case 'a': w = 0, p = ATTOSEC;
                                    break;
                                case 'L': w = 3, p = MILLISEC;
                                    break;
                                case 'U': w = 6, p = MICROSEC;
                                    break;
                                case 'N': w = 9, p = NANOSEC;
                                    break;
                                case 'P': w = 12, p = PICOSEC;
                                    break;
                                case 'F': w = 15, p = FEMTOSEC;
                                    break;
                                case 'A': w = 18, p = ATTOSEC;
                                    break;
                                default: throw "Unrecognized format";
                            }
                            oss.width(w);
                            oss << getFractions(p);
                        }
                    }
                } else {
                    oss << *ptr;
                }
            }

            return oss.str();
        }


        void TimeDuration::fromHash(const karabo::util::Hash& hash) {
            m_Seconds = hash.get<unsigned long long>("seconds");
            m_Fractions = hash.get<unsigned long long>("fractions");
            sanitize(m_Seconds, m_Fractions);
        }


        void TimeDuration::toHash(karabo::util::Hash& hash) {
            hash.set<unsigned long long>("seconds", getSeconds());
            hash.set<unsigned long long>("fractions", getFractions(ATTOSEC));
        }


        TimeDuration& TimeDuration::operator*=(TimeValue factor) {
            // Do not care about overflowing seconds - that's billions of years...
            m_Seconds *= factor;

            const std::pair<TimeValue, TimeValue> fractions = safeMultiply(m_Fractions, factor);
            // Treat non-overflow part...
            m_Fractions = fractions.second % m_oneSecondInAtto;
            m_Seconds += (fractions.second / m_oneSecondInAtto);
            // ... and add overflow if needed:
            if (fractions.first) {
                // Calculate duration of single overflow
                TimeValue attos = TimeValue(0ull) - 1ull; // all bits are set, i.e. maximum TimeValue (which is unsigned)!
                TimeValue seconds = 0ull;
                sanitize(seconds, attos);
                // + 1 since overflow is one more than all 64 bits set
                // (no harm if attos + 1 == 1 second - constructor sanitizes)
                TimeDuration overflow(seconds, attos + 1);
                // Multiply with number of overflows and add
                overflow *= fractions.first; // recursion...
                (*this) += overflow;
            }

            return *this;
        }


        unsigned long long safeAddToFirst(unsigned long long& first, unsigned long long second) {
            // Add 'second' to 'first' and return overflow bits shifted by 64 bit to the right

            unsigned long long overflow = 0ull;

            const unsigned long long allBits = -1; // all bits set!
            if (first > allBits - second) {
                ++overflow; // overflow is only one bit
            }
            first += second; // if overflow, remaining bits are OK

            return overflow;
        }


        std::pair<unsigned long long, unsigned long long>
        safeMultiply(unsigned long long a, unsigned long long b) {
            // Inspired by Norman Ramsey's answer at
            // https://stackoverflow.com/questions/1815367/multiplication-of-large-numbers-how-to-catch-overflow

            const unsigned long long maskHigh = (1ull << 32ull) - 1ull; // only lower 32-bits are set

            const unsigned long long aHigh = a >> 32ull;
            const unsigned long long bHigh = b >> 32ull;
            const unsigned long long aLow = a & maskHigh;
            const unsigned long long bLow = b & maskHigh;

            // Now the result is 2**64 * aHigh * bHigh + 2**32 * (aHigh * bLow + aLow * bHigh) + aLow * bLow,
            // so calculate each term:
            unsigned long long low = aLow * bLow;
            const unsigned long long mid1 = aHigh * bLow;
            const unsigned long long mid2 = aLow * bHigh;
            unsigned long long high = aHigh * bHigh;

            // Add high parts of mid1 and mid2 to high:
            high += (mid1 >> 32ull); // Shifting by 32 bits divides by 2**32 while adding it to high implicitly...
            high += (mid2 >> 32ull); // ...multiplies by 2**64, so overall we multiply by 2**32 as desired.

            // Add low parts of mid1 to low, taking care of overflow:
            const unsigned long long mid1Low = mid1 << 32ull; // Shift 32 bits to multiply with 2**32 as desired.
            high += safeAddToFirst(low, mid1Low);

            // Add low parts of mid2 to low, taking care of overflow:
            const unsigned long long mid2Low = mid2 << 32ull; // Shift 32 bits to multiply with 2**32 as desired.
            high += safeAddToFirst(low, mid2Low);

            return std::make_pair(high, low);
        }
    }
}

File: src/jkarabo/karabo/util/time/TimeDuration.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package karabo.util.time;

import java.math.BigDecimal;
import karabo.util.Hash;

/**
 *
 * @author Sergey Esenov serguei.essenov at xfel.eu
 */
public class TimeDuration {

    private long m_Seconds;
    private long m_Fractions;
    private static String DEFAULT_FORMAT;
    static {
        DEFAULT_FORMAT = "%S.%N";
    }

    public TimeDuration() {
        m_Seconds = 0;
        m_Fractions = 0;
    }

    public TimeDuration(Hash hash) {
        m_Seconds = hash.get("seconds");
        m_Fractions = hash.get("fractions");
    }

    public TimeDuration(long seconds, long fractions) {
        m_Seconds = seconds;
        m_Fractions = fractions;
    }

    public TimeDuration(int days, int hours, int minutes, long seconds, long fractions) {
        m_Seconds = days * TIME_UNITS.DAY.duration() + hours * TIME_UNITS.HOUR.duration() + minutes * TIME_UNITS.MINUTE.duration() + seconds;
        m_Fractions = fractions;
    }

    public TimeDuration(TimeDuration other) {
        this(other.m_Seconds, other.m_Fractions);
    }

    public TimeDuration set(long seconds, long fractions) {
        m_Seconds = seconds;
        m_Fractions = fractions;
        return this;
    }

    public TimeDuration set(int days, int hours, int minutes, long seconds, long fractions) {
        m_Seconds = days * TIME_UNITS.DAY.duration() + hours * TIME_UNITS.HOUR.duration() + minutes * TIME_UNITS.MINUTE.duration() + seconds;
        m_Fractions = fractions;
        return this;
    }

    public boolean isNull() {
        return (m_Seconds == 0) && (m_Fractions == 0);
    }

    public long getDays() {
        return m_Seconds / TIME_UNITS.DAY.duration();
    }

    public long getHours() {
        return (m_Seconds / TIME_UNITS.HOUR.duration()) % 24;

    }

    public long getTotalHours() {
        return m_Seconds / TIME_UNITS.HOUR.duration();
    }

    public long getMinutes() {
        return (m_Seconds / TIME_UNITS.MINUTE.duration()) % 60;
    }

    public long getTotalMinutes() {
        return m_Seconds / TIME_UNITS.MINUTE.duration();
    }

    public long getSeconds() {
        return m_Seconds % 60;
    }

    public long getTotalSeconds() {
        return m_Seconds;
    }

    public long getFractions(TIME_UNITS unit) {
        return m_Fractions / unit.duration();
    }

    public long getFractions() {
        return m_Fractions / TIME_UNITS.NANOSEC.duration();
    }
    
    public String format(String fmt) {
        StringBuilder oss = new StringBuilder();
        //String newLine = System.getProperty("line.separator");
        boolean percent = false;
        for (Character c : fmt.toCharArray()) {
            if (c == '%') {
                if (!percent)
                    percent = true;
                else {
                    percent = false;
                    oss.append('%');
                }
            } else if (percent) {
                percent = false;
                switch(c) {
                    case 'd':
                    case 'D':
                        oss.append(getDays());
                        break;
                    case 'H':
                        oss.append(String.format("%02d", getHours()));
                        break;
                    case 'M':
                        oss.append(String.format("%02d", getMinutes()));
                        break;
                    case 'S':
                        oss.append(String.format("%02d", getSeconds()));
                        break;
                    case 'h':
                        oss.append(String.format("%d", getHours()));
                        break;
                    case 'm':
                        oss.append(String.format("%d", getMinutes()));
                        break;
                    case 's':
                        oss.append(String.format("%d", getSeconds()));
                        break;
                    case 'l':
                        oss.append(String.format("%d", getFractions(TIME_UNITS.MILLISEC)));
                        break;
                    case 'u':
                        oss.append(String.format("%d", getFractions(TIME_UNITS.MICROSEC)));
                        break;
                    case 'n':
                        oss.append(String.format("%d", getFractions(TIME_UNITS.NANOSEC)));
                        break;
                    case 'p':
                        oss.append(String.format("%d", getFractions(TIME_UNITS.PICOSEC)));
                        break;
                    case 'f':
                        oss.append(String.format("%d", getFractions(TIME_UNITS.FEMTOSEC)));
                        break;
                    case 'a':
                        oss.append(String.format("%d", getFractions(TIME_UNITS.ATTOSEC)));
                        break;
                    case 'L':
                        oss.append(String.format("%03d", getFractions(TIME_UNITS.MILLISEC)));
                        break;
                    case 'U':
                        oss.append(String.format("%06d", getFractions(TIME_UNITS.MICROSEC)));
                        break;
                    case 'N':
                        oss.append(String.format("%09d", getFractions(TIME_UNITS.NANOSEC)));
                        break;
                    case 'P':
                        oss.append(String.format("%012d", getFractions(TIME_UNITS.PICOSEC)));
                        break;
                    case 'F':
                        oss.append(String.format("%015d", getFractions(TIME_UNITS.FEMTOSEC)));
                        break;
                    case 'A':
                        oss.append(String.format("%018d", getFractions(TIME_UNITS.ATTOSEC)));
                        break;
                    default:
                        throw new RuntimeException("TimeDuration format is not recognized: " + fmt);
                }
            } else {
                oss.append(c);
            }
        }
        return oss.toString();
    }

    public boolean equals(TimeDuration other) { // operator ==
        return (m_Seconds == other.m_Seconds) && (m_Fractions == other.m_Fractions);
    }

    public long comparesTo(TimeDuration other) {
        if (m_Seconds != other.m_Seconds) {
            return m_Seconds - other.m_Seconds;
        }
        return m_Fractions - other.m_Fractions;
    }

    public boolean isEqual(TimeDuration other) { // operator ==
        return equals(other);
    }

    public boolean isNotEqual(TimeDuration other) {// operator !=
        return !equals(other);
    }

    public boolean isLesser(TimeDuration other) { // operator <
        return comparesTo(other) < 0;
    }

    public boolean isLesserOrEqual(TimeDuration other) { // operator <=
        return comparesTo(other) <= 0;
    }

    public boolean isGreater(TimeDuration other) { // operator >
        return comparesTo(other) > 0;
    }

    public boolean isGreaterOrEqual(TimeDuration other) { // operator >
        return comparesTo(other) >= 0;
    }

    public TimeDuration plusEquals(TimeDuration other) {
        m_Seconds += other.m_Seconds;
        m_Fractions += other.m_Fractions;
        if (m_Fractions > TIME_UNITS.ATTOSEC.duration()) {
            m_Seconds += 1;
            m_Fractions -= TIME_UNITS.ATTOSEC.duration();
        }
        return this;
    }

    public TimeDuration plus(TimeDuration other) {
        TimeDuration result = new TimeDuration(this);
        return result.plusEquals(other);
    }

    public TimeDuration minusEquals(TimeDuration other) {
        m_Seconds -= other.m_Seconds;
        if (m_Fractions < other.m_Fractions) {
            m_Fractions = (TIME_UNITS.ATTOSEC.duration() + m_Fractions) - other.m_Fractions;
            m_Seconds -= 1;
        } else {
            m_Fractions -= other.m_Fractions;
        }
        return this;
    }

    public TimeDuration minus(TimeDuration other) {
        TimeDuration result = new TimeDuration(this);
        return result.minusEquals(other);
    }

    public double divide(TimeDuration other) {
        BigDecimal eighteenth = new BigDecimal("0.000000000000000001");
        BigDecimal thisValue = BigDecimal.valueOf(m_Seconds);
        BigDecimal fraction = BigDecimal.valueOf(m_Fractions);
        thisValue.add(fraction.multiply(eighteenth));
        BigDecimal otherValue = BigDecimal.valueOf(other.m_Seconds);
        BigDecimal otherFraction = BigDecimal.valueOf(other.m_Fractions);
        otherValue.add(otherFraction.multiply(eighteenth));
        return (thisValue.divide(otherValue)).doubleValue();
    }

    public void fromHash(Hash hash) {
        m_Seconds = hash.get("seconds");
        m_Fractions = hash.get("fractions");
    }

    public void toHash(Hash hash) {
        hash.set("seconds", getSeconds());
        hash.set("fractions", getFractions(TIME_UNITS.ATTOSEC));
    }

    public static void setDefaultFormat(String fmt) {
        DEFAULT_FORMAT = fmt;
    }

    @Override
    public String toString() {
        return format(DEFAULT_FORMAT);
    }

    public static String toString(TimeDuration duration) {
        return duration.toString();
    }
}
